plugin/gcp_config.go

```diff
@@ -0,0 +1,113 @@
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/go-gcp-common/gcputil"
+	"github.com/hashicorp/vault/sdk/framework"
+	"google.golang.org/api/compute/v1"
+	"google.golang.org/api/iam/v1"
+)
+
+// gcpConfig contains all config required for the GCP backend.
+type gcpConfig struct {
+	Credentials  *gcputil.GcpCredentials `json:"credentials"`
+	IAMAliasType string                  `json:"iam_alias"`
+	GCEAliasType string                  `json:"gce_alias"`
+}
+
+// standardizedCreds wraps gcputil.GcpCredentials with a type to allow
+// parsing through Google libraries, since the google libraries struct is not
+// exposed.
+type standardizedCreds struct {
+	*gcputil.GcpCredentials
+	CredType string `json:"type"`
+}
+
+const serviceAccountCredsType = "service_account"
+
+// formatAsCredentialJSON converts and marshals the config credentials
+// into a parsable format by Google libraries.
+func (c *gcpConfig) formatAndMarshalCredentials() ([]byte, error) {
+	if c == nil || c.Credentials == nil {
+		return []byte{}, nil
+	}
+
+	return json.Marshal(standardizedCreds{
+		GcpCredentials: c.Credentials,
+		CredType:       serviceAccountCredsType,
+	})
+}
+
+// Update sets gcpConfig values parsed from the FieldData.
+func (c *gcpConfig) Update(d *framework.FieldData) (bool, error) {
+	if d == nil {
+		return false, nil
+	}
+
+	changed := false
+
+	if v, ok := d.GetOk("credentials"); ok {
+		creds, err := gcputil.Credentials(v.(string))
+		if err != nil {
+			return false, errwrap.Wrapf("failed to read credentials: {{err}}", err)
+		}
+
+		if len(creds.PrivateKeyId) == 0 {
+			return false, errors.New("missing private key in credentials")
+		}
+
+		c.Credentials = creds
+		changed = true
+	}
+
+	rawIamAlias, exists := d.GetOk("iam_alias")
+	if exists {
+		iamAlias := rawIamAlias.(string)
+		if iamAlias != c.IAMAliasType {
+			c.IAMAliasType = iamAlias
+			changed = true
+		}
+	}
+
+	rawGceAlias, exists := d.GetOk("gce_alias")
+	if exists {
+		gceAlias := rawGceAlias.(string)
+		if gceAlias != c.GCEAliasType {
+			c.GCEAliasType = gceAlias
+			changed = true
+		}
+	}
+
+	return changed, nil
+}
+
+func (c *gcpConfig) getIAMAlias(role *gcpRole, svcAccount *iam.ServiceAccount) (alias string, err error) {
+	aliasType := c.IAMAliasType
+	if aliasType == "" {
+		aliasType = defaultIAMAlias
+	}
+
+	aliaser, exists := allowedIAMAliases[aliasType]
+	if !exists {
+		return "", fmt.Errorf("invalid IAM alias type: must be one of: %s", strings.Join(allowedIAMAliasesSlice, ", "))
+	}
+	return aliaser(role, svcAccount), nil
+}
+
+func (c *gcpConfig) getGCEAlias(role *gcpRole, instance *compute.Instance) (alias string, err error) {
+	aliasType := c.GCEAliasType
+	if aliasType == "" {
+		aliasType = defaultGCEAlias
+	}
+
+	aliaser, exists := allowedGCEAliases[aliasType]
+	if !exists {
+		return "", fmt.Errorf("invalid GCE alias type: must be one of: %s", strings.Join(allowedIAMAliasesSlice, ", "))
```

Comment:
Wondering if the `strings.Join()` here should use `allowedGCEAliasesSlice` instead of `allowedIAMAliasesSlice`?  Looks like `allowedGCEAliasesSlice` is an unused variable.

Reply:
Good catch! Copy-pasta mistake. Fixed.